bindings/pyroot_experimental/cppyy/CPyCppyy/src/CPPInstance.h

```diff
@@ -0,0 +1,113 @@
+#ifndef CPYCPPYY_CPPINSTANCE_H
+#define CPYCPPYY_CPPINSTANCE_H
+
+//////////////////////////////////////////////////////////////////////////////
+//                                                                          //
+// CpyCppyy::CPPInstance                                                    //
+//                                                                          //
+// Python-side proxy, encapsulaties a C++ object.                           //
+//                                                                          //
+//////////////////////////////////////////////////////////////////////////////
+
+
+// Bindings
+#include "CPPScope.h"
+#include "Cppyy.h"
+#include "CallContext.h"     // for Parameter
+
+
+// TODO: have an CPPInstance derived or alternative type for smart pointers
+
+namespace CPyCppyy {
+
+class CPPInstance {
+public:
+    enum EFlags {
+        kNone        = 0x0,
+        kIsOwner     = 0x0001,
+        kIsReference = 0x0002,
+        kIsRValue    = 0x0004,
+        kIsValue     = 0x0008,
+        kIsSmartPtr  = 0x0010,
+        kIsPtrPtr    = 0x0020 };
+
+public:
+    void Set(void* address, EFlags flags = kNone)
+    {
+    // Initialize the proxy with the pointer value 'address.'
+        fObject = address;
+        fFlags  = flags;
+        fSmartPtrType = (Cppyy::TCppType_t)0;
+        fDereferencer = (Cppyy::TCppMethod_t)0;
+    }
+
+    void SetSmartPtr(Cppyy::TCppType_t ptrtype, Cppyy::TCppMethod_t deref)
+    {
+        fFlags |= kIsSmartPtr;
+        fSmartPtrType = ptrtype;
+        fDereferencer = deref;
+    }
+
+    void* GetObject() const
+    {
+    // Retrieve a pointer to the held C++ object.
```

Comment:
Could you make this proper doxygen doc? Even if it becomes an upstream patch ;-)

Reply:
When you say "this" you mean this file in particular or also the rest of the 60 files in CPyCppyy that do not follow doxygen conventions? :smile: This requires quite some effort.

Reply:
Ignore my comment (I thought this was part of `PyROOT/`)